### DISCUSSION TITLE
add_requires 支持git地址

### DISCUSSION THREAD

**Discussion by featherfly** [xmake-io/xmake#2977] [Ideas]
### 你在什么场景下需要该功能？

我想直接把我在git仓库的lib拉去下来
就像platformio的包管理那样，既可以使用名称拉取已经在中央库注册的，也可以直接使用git url拉取

### 描述可能的解决方案

add_requires("https://github.com/xxx/yyy.git")

**Comment by waruqi** [comment]
不考虑，c++ 的每个库都会有各种复杂的编译处理，构建系统也是多样化的，有可能还得加 patch，除非很简单的 headeronly 库，大部分都要做特殊处理，你随便到 xmake-repo 里面仓库包看下，基本不会有只需要一个 git url，什么都不配置，就能完全搞定的，https://github.com/xmake-io/xmake-repo/blob/master/packages/p/pcre2/xmake.lua

如果是复杂的 c++ 库，不太可能实现，如果是简单的库，直接定义包 加 git url 就行了，也没几行代码。

```lua
package("test")
    set_urls("xxxx.git")
    on_install(function (package)
    end)
package_end()

add_requires("test")
```

**Comment by featherfly** [comment]
按照这个
package("test")
    set_urls("xxxx.git")
    on_install(function (package)
    end)
package_end()

add_requires("test")

编译找不到头文件

**Reply by waruqi** [reply]
看文档和 xmake-repo 里面例子 https://xmake.io/#/zh-cn/package/remote_package?id=%e5%8c%85%e6%8f%8f%e8%bf%b0%e8%af%b4%e6%98%8e

**Reply by featherfly** [reply]
还是找不到头文件，我已经在安装目录的include目录看到头文件了
环境 windows11 mingw x86_64
安装目录 .xmake\packages\c\c_crc\master\664afa6891ed459e9a6c3e9948b6e651\include\crc8.h

> xmake.lua

```
package("c_crc")
    set_urls("https://gitee.com/featherfly/c_crc.git")
    on_install(function (package)
        local configs = {}
        if package:config("shared") then
            configs.kind = "shared"
        end
        os.cp("include/*.h", package:installdir("include"))
        import("package.tools.xmake").install(package, configs)
    end)
package_end()

add_requires("c_crc")

add_rules("mode.debug", "mode.release")
target("xmake-repo-test")
    set_kind("binary")
    add_files("src/*.c")

```

> xmake --verbose
```
[ 25%]: cache compiling.release src\main.c
Y:\soft\programe\runtime\msys64\mingw64\bin\x86_64-w64-mingw32-gcc -c -m64 -fvisibility=hidden -O3 -DNDEBUG -o build\.objs\xmake-repo-test\mingw\x86_64\release\src\main.c.obj src\main.c
error: src\main.c:3:10: fatal error: crc8.h: No such file or directory
    3 | #include "crc8.h"
      |          ^~~~~~~~
compilation terminated.
```

编译没有 -I 参数引用add_requires的包

**Reply by waruqi** [reply]
仔细看文档 https://xmake.io/#/zh-cn/package/remote_package

**Reply by featherfly** [reply]
感谢作者的回答，终于搞定了

我有个问题，我在tbox的构建文件中看到了set_kind("$(kind)")
我自己试着这样写就安装有问题，只有改为set_kind("static")才能正常
我现在想在引入指定库的时候决定构建的类型（static, object)等，应该如何配置
文档中的shared=True可以决定是静态编译还是动态编译
我想使用kind object编译库，并只链接其中的一部分应该如何配置

例如我有个crc库，里面有crc8 crc16 crc32等实现，我现在使用了crc8，只想要链接crc8而不是链接整个库
我应该如何实现，包括crc库的发布以及crc库的引用

**Reply by waruqi** [reply]
> 我有个问题，我在tbox的构建文件中看到了set_kind("$(kind)")

就这么用，有问题自己根据错误排查下，确认符号是否导出

> 我现在想在引入指定库的时候决定构建的类型（static, object)等，应该如何配置

add_requires("xxx", {configs = {shared = true}})

>         os.cp("include/*.h", package:installdir("include"))

add_headerfiles("include/*.h") in xmake.lua

**Reply by featherfly** [reply]
我知道add_requires("xxx", {configs = {shared = true}})可以编译动态库

 我现在想要把引入的库，编译成object，应该如何配置，我试着在on_install里设置configs.kind = "object" ，但是安装后lib目录为空

**Reply by waruqi** [reply]
库目前只能作为 static/shared 引入，不支持 object 引入，从没见这么用的。。

**Comment by featherfly** [comment]
为了发布一个基础最基础的lib（不包含依赖，例如crc8这样就一个.h一个.c文件的库），按照文档走到私服这步，然后就一直test不过，已经折腾了一天了
我用过maven（gradle）,npm, cargo, platformio 这么多能够进行包管理和发布的工具，xmake是第一个让我连一个基础库都搞不定的

**Reply by waruqi** [reply]
> 我用过maven（gradle）,npm, cargo, platformio 这么多能够进行包管理和发布的工具，xmake是第一个让我连一个基础库都搞不定的

呵，拿 maven/npm/cargo 去跟 c++ 的包管理比，我也是第一次看到这么对比的，也是醉了。。要是搞 js java rust 当然没 c++ 那么多破事，不用处理平台差异，编译器差异，构建工具差异

你大可以试试 c++ 的其他包管理，conan/vcpkg 去自定义个包试试

**Reply by waruqi** [reply]
> 为了发布一个基础最基础的lib（不包含依赖，例如crc8这样就一个.h一个.c文件的库），按照文档走到私服这步，然后就一直test不过，已经折腾了一天了

自己多看看文档，多看看 xmake-repo 里面其他的包 

https://github.com/xmake-io/xmake-repo/blob/f3c03c4c51a6086eb34b38f66ad2d5bbfcba0eac/packages/q/qoi/xmake.lua

这种，也没见哪个其他用户能折腾一天的。。就那么几行配置。

**Reply by featherfly** [reply]
platformio也是c c++，我发布库也是跟着文档走下去就搞定了
差异其实是在文档上，xmake的文档的教程属性太差了
作者不要上头，我不是贬低xmake，因为c的其他构建工具更难用
我现在折腾xmake就是想把我这里写嵌入式的内容都用xmake管理起来
